### PR TITLE
make spring starter stable

### PR DESCRIPTION
**Pull request by zeitlinger**

### CODE REVIEW

**Review by trask — COMMENTED** [review]
we need to add `otel.japicmp-conventions` to the stable modules, see #7046

I'm not sure why we don't just add `otel.japicmp-conventions` centrally, see https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-java%20otel.japicmp-conventions&type=code (compared to how we are currently doing it in our repo)

File: instrumentation/spring/spring-boot-autoconfigure-3/build.gradle.kts [hunk 1]
```diff
@@ -13,7 +13,7 @@ otelJava {
 dependencies {
   val springBootVersion = "3.2.4"
   library("org.springframework.boot:spring-boot-starter-web:$springBootVersion")
-  compileOnly(project(":instrumentation:spring:spring-boot-autoconfigure"))
+  implementation(project(":instrumentation:spring:spring-boot-autoconfigure"))
```
Comment (trask) [inline instrumentation/spring/spring-boot-autoconfigure-3/build.gradle.kts:16]: can you send this and the other non-stability change below (additional-spring-configuration-metadata.json) as a separate PR? just mark it with the milestone and I'll make sure it gets into the release

Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure-3/build.gradle.kts:16]: sure - https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/11784

**Comment by zeitlinger** [comment]
> we need to add `otel.japicmp-conventions` to the stable modules, see #7046
> 
> I'm not sure why we don't just add `otel.japicmp-conventions` centrally, see https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-java%20otel.japicmp-conventions&type=code (compared to how we are currently doing it in our repo)

maybe because we're publishing things we don't want to have api diffs because they are not really libraries?

**Comment by trask** [comment]
@zeitlinger can you resolve the new merge conflict? thanks

File: docs/apidiffs/current_vs_latest/opentelemetry-spring-boot.txt [hunk 2]
```diff
@@ -0,0 +1,137 @@
+Comparing source compatibility of opentelemetry-spring-boot-2.6.0-SNAPSHOT.jar against 
++++  NEW ANNOTATION: PUBLIC(+) ABSTRACT(+) io.opentelemetry.instrumentation.spring.autoconfigure.EnableOpenTelemetry  (not serializable)
```
Comment (trask) [inline docs/apidiffs/current_vs_latest/opentelemetry-spring-boot.txt:2]: I couldn't find any usage of this annotation, is there some magic involved? thanks

File: conventions/src/main/kotlin/otel.japicmp-conventions.gradle.kts [hunk 3]
```diff
@@ -91,7 +91,7 @@ if (project.findProperty("otel.stable") == "true") {
 
         // this is needed so that we only consider the current artifact, and not dependencies
         ignoreMissingClasses.set(true)
-        packageExcludes.addAll("*.internal", "*.internal.*")
+        packageExcludes.addAll("*.internal", "*.internal.*", "*.shaded", "*.shaded.*")
```
Comment (trask) [inline conventions/src/main/kotlin/otel.japicmp-conventions.gradle.kts:94]: should we update https://github.com/open-telemetry/opentelemetry-java/blob/main/VERSIONING.md? (and/or our repo copy: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/VERSIONING.md):

> A backwards-incompatible change affects the public API of a module. The public API is any public
class or method that is not in a package which includes the word `internal`.

or maybe we should put the shaded classes under `.internal.shaded`?

Comment (zeitlinger) [reply conventions/src/main/kotlin/otel.japicmp-conventions.gradle.kts:94]: `.internal.shaded` seems to be the easiest

File: docs/apidiffs/current_vs_latest/opentelemetry-spring-boot.txt (hunk 2 above)
Comment (zeitlinger) [reply docs/apidiffs/current_vs_latest/opentelemetry-spring-boot.txt:2]: yes, `@ComponentScan(basePackages = "io.opentelemetry.instrumentation.spring.autoconfigure")` is the magic

@jeanbisutti I think we didn't document this anywhere

**Review by trask — APPROVED** [review]

**Pull request merged**